distribution/lib/Standard/Base/0.0.0-dev/src/Panic.enso

```diff
@@ -0,0 +1,268 @@
+                            False -> Panic.throw caught_panic
+                    _ -> Panic.throw caught_panic
+
+    ## Executes the provided action and if a Java exception matching the provided type was
+       thrown, calls the provided callback.
+
+       Normally, Java exceptions are wrapped in a `Polyglot_Error` instance, so
+       using a `Panic.catch` requires unwrapping the error by calling
+       `caught_panic.payload.cause`. This helper function allows the handler to
+       work with the Java exception directly. The downside is that if the Java
+       exception is rethrown, it will be rethrown as a Java exception object
+       wrapped in an Enso panic. So if the handler needs to rethrow the original
+       exception preserving its shape and stacktrace, `Panic.catch` should still
+       be preferred.`
+
+       > Example
+         Convert a string to an integer, catching the Java `NumberFormatException`
+         and converting it to a more Enso-friendly dataflow error.
+
+             polyglot java import java.lang.Long
+             polyglot java import java.lang.NumberFormatException
+             parse str =
+                 Panic.catch_java NumberFormatException (Long.parseLong str) java_exception->
+                     Error.throw (Illegal_Argument.Error "The provided string is not a valid number: "+java_exception.getMessage)
+    catch_java : Any -> Any -> (Throwable -> Any) -> Any
+    catch_java panic_type ~action handler =
+        Panic.catch_primitive action caught_panic-> case caught_panic.payload of
+            Polyglot_Error.Polyglot_Error_Data java_exception ->
+                case (panic_type == Any) || (java_exception.is_a panic_type) of
+                    True -> handler java_exception
+                    False -> Panic.throw caught_panic
+            _ -> Panic.throw caught_panic
+
+    ## Executes the provided action and converts a possible panic matching any of
+       the provided types into a dataflow Error.
+
+       If action executes successfully, the result of `Panic.recover` is the result
+       of that action. Otherwise, if it panicked with a type matching one of the
+       expected error types, that panic is returned as a dataflow error. Unexpected
+       panics are passed through as-is. it is the panic that was thrown after
+       conversion to a dataflow error.
+
+       Arguments:
+       - expected_types: The types of expected panics which should be recovered.
+         This can either be a Vector of types or a single type.
+       - action: The code to execute that potentially panics.
+
+       > Example
+         Converting an expected panic to a dataflow error.
+
+             Panic.recover Illegal_Argument.Error (Panic.throw (Illegal_Argument.Error "Oh!" Nothing))
+
+       > Example
+         Converting one of many expected panic types to a dataflow error.
+
+             Panic.recover [Illegal_Argument, Illegal_State] (Panic.throw (Illegal_Argument.Error "Oh!" Nothing))
```

Comment:
I'm afraid that this example will not work as shown here. See the discussion above.
```suggestion
             Panic.recover [Illegal_Argument.Error, Illegal_State.Error] (Panic.throw (Illegal_Argument.Error "Oh!" Nothing))
```

I think we really should fix `is_a` as indeed `(Illegal_Argument_Error_Data "Foo").is_a Illegal_Argument_Error` **should** be `True` but currently it is not.